Add codex-rs --dangerously-run-with-no-sandbox

cliff-openai opened the pull request:
This should address #1254 (and motivation/justification is available there)

In short, the sandbox, even with network+disk access, still blocks some things on osx (like access to the keychain most likely). Also, in some environments people are already fully sandboxed (externally to codex), so this gives an option in those environments.

cliff-openai committed:
Add codex-rs --dangerously-auto-approve-everything. Fix #1254

cliff-openai commented:
I have read the CLA Document and I hereby sign the CLA

bolinfest commented:
Given that this fails for me today:

```
$ codex debug seatbelt -s disk-full-read-access -s disk-full-write-access -s network-full-access -- gh issue list --limit 10
Post "https://api.github.com/graphql": tls: failed to verify certificate: x509: OSStatus -26276
```

I don't think this quite accomplishes the goal (or at least, I'm not sure it matches user expectations). It's a little tricky because the are two dimensions to "permissiveness" that can be configured independently:

- which commands are auto-approved
- what sandbox configuration to apply when running a command

Note the latter is determined by `enum SandboxType` and I suspect users expect `--dangerously-auto-approve-everything` to correspond to `SandboxType::None`. Or at least, I believe `SandboxType::None` needs to be used for Codex to be able to run `gh issue list --limit 10`.

https://github.com/openai/codex/blob/b73426c1c40187ca13c74c03912a681072c2884f/codex-rs/core/src/exec.rs#L65-L74

To that end, I don't know if `--dangerously-auto-approve-everything` is quite the right name anymore because it would not just be relaxing approvals, but sandbox constraints, as well.

For reference, my personal `config.toml` has:

```
approval_policy = "never"

sandbox_permissions = [
    "disk-full-read-access",
    "disk-write-platform-user-temp-folder",
    "disk-write-platform-global-temp-folder",
    "disk-write-cwd",
    "disk-write-folder=/Users/mbolin/.pyenv/shims",
]
```

In other words, this already satisfies the case where I don't want to be prompted for anything, but I do want it everything to be run under a sandbox policy that is roughly equivalent to `--full-auto`.

cliff-openai committed:
rename option to --dangerously-auto-approve-and-run-without-sandbox

cliff-openai commented:
Ok, I renamed the option to `--dangerously-auto-approve-and-run-without-sandbox`. If you would prefer, I could instead make the option just be `--dangerously-run-commands-without-sandbox` and then users may also need/want to set `-a never`?

bolinfest commented:
https://github.com/openai/codex/issues/1248 is somewhat related in that I am not 100% happy with the contract I implemented for `--sandbox-permission` today.

bolinfest commented:
@cliff-openai though given this investigation, would using the following config unblock you for the moment?

```
sandbox_permissions = [
    "disk-full-read-access",
    "disk-full-write-access",
    "network-full-access",
]
```

Though assuming this works, this feels like a bug because it should be possible to use this in concert with `approval_policy = "unless-allow-listed"`, which should mean, "Always ask me, but if I approve, do not run the command in a sandbox," but based on the code above, it appears that if `is_unrestricted()` is `true`, then `AskForApproval` is not considered.

oai-ragona commented on codex-rs/core/src/protocol.rs:
```diff
@@ -176,6 +176,18 @@ impl SandboxPolicy {
 
+    /// Unrestricted policy: allow full disk read/write and network access.
+    /// Commands will be auto-approved without sandbox enforcement.
+    pub fn new_unrestricted_policy() -> Self {
```
I think this will still end up invoking Landlock/Seccomp on linux. In the actual `exec.rs` code I think we want a conditional to just straight up not sandbox at all to achieve the properties you're looking for here.

cliff-openai replied:
ok, I think that the current version now does this (more explicitly has a "NoSandbox" implementation)

cliff-openai committed:
Change to --dangerously-run-with-no-sandbox

This option is now explicitly just about sandbox policy, and we are
explicit/clear about the "NoSandbox" state.

cliff-openai commented:
ok, I've updated the change to be more explicitly just about creating a "no sandbox" option.

I am not blocked, and am happy to just abandon this change if it is not helpful. I was assuming that the presence of issue #1254 meant that this change was wanted, but no worries if not.

cliff-openai committed:
cargo fmt
